### PR TITLE
core: fix a race in ServerImpl for multi-transport-server

### PR DESCRIPTION
**Pull request by dapengzhang0**
As discovered in #6601 and its partial fix #6610,  `ServerImpl` (https://github.com/grpc/grpc-java/blob/v1.26.0/core/src/main/java/io/grpc/internal/ServerImpl.java#L184) and  `NettyServer` (https://github.com/grpc/grpc-java/blob/v1.26.0/netty/src/main/java/io/grpc/netty/NettyServer.java#L251) have risk of deadlock.

Targeting the issue #6641.

There might be two things need be done to make the code healthy:
1. move `ts.start(listener)` out of synchronization block
2. move multiple port support from `ServerImpl` to `NettyServer`.

This PR is working on (1).

### CODE REVIEW

File: core/src/test/java/io/grpc/internal/ServerImplTest.java [hunk 1]
```diff
@@ -420,7 +420,6 @@ public void start(ServerListener listener) throws IOException {
     } catch (IOException e) {
       assertSame(ex, e);
     }
-    verifyNoMoreInteractions(executorPool);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/ServerImplTest.java]: Why was this removed?

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 2]
```diff
@@ -388,6 +409,7 @@ public ServerTransportListener transportCreated(ServerTransport transport) {
     public void serverShutdown() {
       ArrayList<ServerTransport> copiedTransports;
       Status shutdownNowStatusCopy;
+      awaitStartComplete();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/ServerImpl.java]: This looks dangerous.

I'd much rather we set `activeTransportServers = transportServers.size()` within the initial synchronization block of start() to prevent needing to block here.

Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ServerImpl.java]: set `activeTransportServers = transportServers.size()` is broken if one of `ts.start()` throws because not enough times of `listern.serverShutdown()` will be called .

File: core/src/main/java/io/grpc/internal/ServerImpl.java [hunk 3]
```diff
@@ -269,13 +277,26 @@ public ServerImpl shutdown() {
...
         ts.shutdown();
       }
     }
     return this;
   }
 
+  private void awaitStartComplete() {
+    boolean startCompleted = false;
+    try {
+      startCompleted = startComplete.await(30, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ServerImpl.java]: We can assert that is true.

File: core/src/test/java/io/grpc/internal/ServerImplTest.java (hunk 1 above)
Comment (dapengzhang0) [reply core/src/test/java/io/grpc/internal/ServerImplTest.java]: Because `executorPool.getObject()` is called after the change.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 2 above)
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ServerImpl.java]: Seems we need redesign multiple transportServers support in future. It's just too inconvenient now. For now the fix seems working. Can you elaborate what is the danger?

**Comment by dapengzhang0** [comment]
> Hmmm... It is really awkward that we can have bind succeed for one transport and have it fail for another and then start() with throw and leave things in a half-started state. Previously, when there was only one transport, ts.start() throwing would leave started == false.

Yes previously it was broken in multiple ways.

File: core/src/main/java/io/grpc/internal/ServerImpl.java (hunk 2 above)
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/ServerImpl.java]: Now I see the danger, because `serverShutdown()` is running in the event loop, if another `ts.start()` submits something into the event loop while `awaitStartComplete()` in  `serverShutdown()`, then it will be deadlock again.

**Comment by ejona86** [comment]
We gave up on trying to fix Server for multiple ports. The API is just wrong to support partial failures. The multiple-port handling will be moved into the Netty transport, where partial failures can be handled.

**Comment by dapengzhang0** [comment]
Close in favor of #7674 and subsequent PRs
